geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/sortedset/ZRangeByScoreExecutor.java

```diff
@@ -33,35 +30,35 @@
 public class ZRangeByScoreExecutor extends AbstractExecutor {
   @Override
   public RedisResponse executeCommand(Command command, ExecutionHandlerContext context) {
-    RedisSortedSetCommands redisSortedSetCommands = context.getSortedSetCommands();
-
     List<byte[]> commandElements = command.getProcessedCommand();
 
-    SortedSetRangeOptions rangeOptions;
-    boolean withScores = false;
+    SortedSetScoreRangeOptions rangeOptions;
 
     try {
       byte[] minBytes = commandElements.get(2);
       byte[] maxBytes = commandElements.get(3);
-      rangeOptions = new SortedSetRangeOptions(minBytes, maxBytes);
+      rangeOptions = new SortedSetScoreRangeOptions(minBytes, maxBytes);
     } catch (NumberFormatException ex) {
       return RedisResponse.error(ERROR_MIN_MAX_NOT_A_FLOAT);
     }
 
     // Native redis allows multiple "withscores" and "limit ? ?" clauses; the last "limit"
     // clause overrides any previous ones
-    if (commandElements.size() >= 5) {
-      int currentCommandElement = 4;
+    // In order to match the error reporting behaviour of native Redis, the argument list must be
+    // parsed in reverse order. Stop parsing at index = 4, since 0 is the command name, 1 is the
+    // key, 2 is the min and 3 is the max
+    boolean withScores = false;
 
-      while (currentCommandElement < commandElements.size()) {
+    if (commandElements.size() >= 5) {
```

Comment:
This whole block of code is much the same as the one in ZRangeByLexExecutor. I think all you need is somehow to tell rangeOptions.parseLimitArguments that it should also parse the withScores flag. You could then ask rangeOptions if isWithScores is true.
You could have the method that does this whole for loop in rangeOptions in a method that returns a RedisResponse. If it returns one then this code would return that response. Otherwise it would continue

Reply:
I plan to do a major refactor of this code as part of the ZREVRANGEBYLEX work, once all the other range commands are implemented, as you're right, there's a lot of duplication and potential for simplification. I think we could probably manage to have a single AbstractZRangeExecutor class, extended by the ZRange, ZRevRange, ZRangeBYScore, ZRevRangeBYScore, ZRangeByLex and ZRevRangeByLex executors that populates an AbstractSortedSetRangeOptions class, extended by SortedSetRankRangeOptions, SortedSetScoreRangeOptions and SortedSetLexRangeOptions classes. Maybe I'm overthinking it though.

Reply:
sounds good